### ISSUE TITLE
[ISSUE] API Cache Damage

### ISSUE THREAD

**Issue by iWARR** [ghacksuserjs/ghacks-user.js#925]
I don't know a better place to ask (may be this relative to some prefs)

I have one big significant and unresolved issue since FF70 (may be even earlier) to a present time.
This is broken API Cache and slowdowns when FF loads 1-st webpage after start (freshy configs, many versions). Note, that FF75 loads first webpage faster than previous versions.

Probably, this is relative to "Temporary containers" Add-on or traditional problems with FF Local storage design. Or, may be, some prefs...

At last, the issue was opened [here](https://github.com/stoically/temporary-containers/issues/393), but still no response.

1) Can you, guys, (who use "Temporary containers") make a glitch test and say if it was passed successfully?
2) Any suggestions are welcome.

**Comment by earthlng** [comment]
the Cache API is the offline cache which most people will never need/use and very few sites use anyway. This user.js disables the offline cache in 2730. Do you not have/use that pref?

Due to that pref being disabled, when I run that testsite with TC it just says "Cache API: Bad: Totally Broken. (fullyBroken)".

I've never noticed any slowdowns or similar problems

**Comment by iWARR** [comment]
I have these:

```
// 2730 : Offline cache for offline browser = [ Disabled ]
/*		Web apps can store data for retrieval, even when the browser is offline.
		JavaScript can acts with them until the connection is restored. 
		If disabled, URLs will not be downloaded for the offline cache.
	[1]	http://kb.mozillazine.org/Browser.cache.offline.enable     */
user_pref("browser.cache.offline.enable", false);			// gHacks //
user_pref("browser.cache.offline.storage.enable", false);		// added // 
user_pref("browser.cache.offline.capacity", 0);				// added //
```

**Comment by iWARR** [comment]
I'm expected a big slowdowns before, even with loading google.com (long time to open) after starting FF. Some other webpages I opened in tabs in the prevous session may stuck until manual refresh. I was very annoing.

This is a fact (for me) for FF70-74. As I said, FF75 demonstrates much better behaviour with this, but I feel not everythig is good. And finally I discover a TC problem. That's all I can say according to my own experience.

Do you recommend to play with resetting (2730) + TC?

**Comment by iWARR** [comment]
In fact, the broken local cache repository (you say "people will never need/use and very few sites use anyway") somehow have influence on webpage loading. When TC breaks the cache, even when uninstalled, the simple loading of test-page may stuck or open very long time.

I think, this is some kind of a "confirmation" that these things are somehow associated...

**Comment by earthlng** [comment]
I'd expect it to also say "Cache API: Bad: Totally Broken. (fullyBroken)". if you have those prefs set. Are you sure your prefs are applied correctly and you don't have a syntax error in your user.js?

**Comment by iWARR** [comment]
Everything I checked (a lot) ever, always look correspondingly inside the about:config
But yeah, it's a big file...

**Comment by crssi** [comment]
Here is says: `Totally Working. (fullyOperational)`

On my profile you can find the configuration I use.
Cheers

**Comment by Gitoffthelawn** [comment]
@iWARR How do you define slowdown?  3 seconds?  20 seconds?  More?

I ran the test you provided, and it took about 3 seconds to run with TC installed and enabled.

Everything in the test reported "Totally Working. (fullyOperational)".

I have Offline Cache enabled.  Does anyone have any links as to why Offline Cache should be disabled, or at least a well-written document explaining what function(s) it serves?

**Comment by earthlng** [comment]
Hmmm, that's weird. Apparently it's got nothing to do with offline cache after all. Sorry bout that.

I have `dom.caches.enabled=false` in my user.js and that alone makes it say "Totally Broken".
Changing that to true and allowing the XHR requests in uMatrix results in "Totally Working".

I thought `dom.caches.enabled` was just for serviceworkers cache but apparently it's more than that.

**Comment by earthlng** [comment]
> Apparently it's got nothing to do with offline cache after all.

or at least if "Cache API" (=appCache?) *does* have something to do with "offline cache" then it's definitely not controlled by any of the `browser.cache.offline.*` prefs.

@iWARR what value do you have in about:config for `dom.storage.next_gen` ? If it's true then maybe that's the reason for some or all of your storage-related and/or slowdown problems.

**Comment by iWARR** [comment]
@Gitoffthelawn 

Even 5-10 sec is too much nowadays for the browser. Opera never has such lags 10 years ago.
Or "I feel" it's stuck for some time, or other tabs may stuck "physically" - **endless loading until manual refresh.** 

And I really don't see the speed profit of the ANY caching Mozilla's tech (when cache is ON or OFF) + I don't feel the profit of putting things to a memory (I have a lot of RAM).

@earthlng 

That's is even MORE intersting, because I'm always keep `dom.caches.enabled` by default (enabled) in case I far long time ago warn myself do not touch this ever. People have too much problems flipping it. And I can read the words.

`dom.storage.next_gen` I've tried in FF70 era. No luck, of course. I keep this also by default (disabled) until Mozilla change its own mind and it will be ready for the "train" trip.

> "Cache API" (=appCache?) does have something to do with "offline cache" then it's definitely not controlled by any of the browser.cache.offline.* prefs.

I feel the same. (And I'm confused).

**Comment by iWARR** [comment]
During the tests in the past, (when I had this lag) I tried to see what will happen if I exclude some things that need I/O operations. The first thing was uBO, because it store a blocking lists. And I was right - the lags **almost** gone. The pages start loading significally faster. But. We need uBO and don't need lags... For example, Opera didn't has such lags on the same uBO version.

So, I think the problem is not in the cache itself, but in the **Local storage** malfunctions. That's why my TC issue name includes : "[ISSUE] TC breaks Cache API **(Local storage)**"

I beleive, TC just demonstrate us the bug in "real deal", but main problem is on Mozilla's side.

**Comment by gwarser** [comment]
https://www.reddit.com/r/uBlockOrigin/comments/d4zlfi/ublock_on_firefox_preventing_pages_from_loading/f0i4nss/

**Comment by iWARR** [comment]
@gwarser

Seems I'm already read this before. I'm sure I tried playng with uBO settings also.
At least, manipulating with uBO settings does zero result.

**Comment by Thorin-Oakenpants** [comment]
> > Apparently it's got nothing to do with offline cache after all.
> 
> or at least if "Cache API" (=appCache?) _does_ have something to do with "offline cache" then it's definitely not controlled by any of the `browser.cache.offline.*` prefs.

Correct: they are not talking about appCache : tested in a nilla profile by flipping `browser.cache.offline.enable` = `false` (restarting for good measure) and re-testing. They've already killed part of this off but left `"applicationCache" in window` (can't be arsed finding the info right now, and it's irrelevant)

Looked at the [source code](view-source:https://firefox-storage-test.glitch.me/script.js), they're not using service workers: so it has nothing to do with sw cache either.

So maybe it's the actual cache (1001) ... so I flipped `browser.cache.disk.enable` = `false` (restarting for good measure) and re-tested - but it's not that

So I'm not sure exactly what they mean by "Cache API" - but the source code reveals more when you read about opaque padding and caches.sqlite

```
/**
 * The Cache API upgrade we're concerned about is from when we upgraded the
 * schema to capture opaque padding.  So we go a little overboard with
 * permutations related to this, especially given that we know that all
 * caches are stored in a single database file.  (Unlike IndexedDB, which
 * stores each content-created database in a separate SQLite database.)
 *
 * Note that when we talk about opaque padding:
 * - It's not directly observable; you need to use navigator.storage.estimate()
 *   to try and figure out if it's there or not.
 * - navigator.storage is only available in nightly...
 * - We don't actually care if it's there or not in these tests, we just want
 *   to make sure that the Response object still "works" in the sense that
 *   cache.match() returns a Response object.
 * - We make assumptions about whether there's padding magic based on the
 *   current Firefox version number, and we use that to decide what cache to
 *   put the response into so we know what we're trying to match() against.
 */

   // snip
  let cacheNames;
    // keys() will trigger full initialization of caches.sqlite if it doesn't
    // exist.

```

**Comment by Thorin-Oakenpants** [comment]
In my FF75 with user.js etc, I get Cache API is all good: but only after I configured uMatrix/uBO properly (and allowed at least session only cookies), otherwise it failed. @iWARR - make sure your tests don't include extensions fucking up your results

Also, how is this a ghacks user.js problem, when IWARR is using his own user.js - I'm tempted to close this as invalid .. but it is interesting (the cache api, not the slowdown)

**Comment by Thorin-Oakenpants** [comment]
> [appCache]....
> What's wrong with just keeping it enabled, and clearing it automatically when Firefox closes?

Nothing, except it will only be cleared at the end of each session. Since 0.0000000002 % of websites use it, and it's been announced as slated for deprecation for about 100 years, it's just one more piece of persistent local web storage that can be eliminated as a complete waste of space. It's also been neglected - i.e bugs with it have been closed as wontfix, and it (from memory) won't show in a lot of places (e.g. manage data, inspector) and don't be surprised if things like forget this site fail to do anything with it, or if it's covered under quota management etc. It's also been neutered since 71 (I think, at least in non stable: can't be bothered digging up the tickets)

**Comment by iWARR** [comment]
When I reset all these 3 prefs to defaults and retest, there is no diffs for the issue:
```
"browser.cache.offline.enable"
"browser.cache.offline.storage.enable"
"browser.cache.offline.capacity"
```
1) Before TC installation - `API Cache : Good`
2) As soon as I install TC and turn on the checkbox `[•] Automatic mode`, then `API Cache : Bad`

Cleaning should be as described on TC thread : 

[Wipe local storage and reset it to default] + Uninstall TC (additionally : delete "permanent" и "temporary" filders inside the profile)

These are naked facts. 10 times tested before posting this and once again now.

**Comment by crssi** [comment]
^^ did you use ghacksuser.js or plain vanilla profile?

**Comment by iWARR** [comment]
> did you use ghacksuser.js or plain vanilla profile?

With TC? No. Probably, I should try...

**Comment by crssi** [comment]
^^ I asked beacuse I tried to replicate your problem by your STR (steps to reproduce) here with a clean new plain vanilla profile and I cannot. It works always as it should.
Can you replicate on another computer?

**Comment by iWARR** [comment]
Bad Cache can be replicated with **my** FF setup.
I had an issue on Win8.1 (Desktop) and Win10 (Notebook).

Anyway, your suggestion about default ghacksuser.js worked and this is step to forward.
Now it's my part - if I'll be lucky, I'll find the guilty bad-boy fast.
Or I'll start excluding one by one (this is long way trip).

**Comment by crssi** [comment]
> Interesting, that test shows me now everywhere `68` instead of `75`, for example:

`68` is fine due to RFP, see:

``` js
/* 4501  */ user_pref("privacy.resistFingerprinting", true);
```

**Comment by iWARR** [comment]
I don't use RFP, but usinig other extra FF prefs + Addons. Just my decision. I'm "Non-RFP" user.
I decide do not make dupes, because Add-ons doing the same in better way.

**Comment by crssi** [comment]
>I don't use RFP... because Add-ons doing the same in better way

I lost you here... its just a comment, do not answer. 😄

**Comment by iWARR** [comment]
It's OK. Mine is too, "just a comment". Any notes from the pro-users are useful for me.

**Comment by iWARR** [comment]
When I complete my future tests, I'll post here, who was quilty.

**Comment by iWARR** [comment]
### THE CRITICAL PREF WAS FOUND

```
// 4001 : First Party Isolation (FPI) = [ Enabled ]
/*		Restricts Cookies, Cache and other data access to the Domain Level (TOR feature).
		So that only the domain that dropped the cookie can access it.
		Highly effective and wide user protection against marketing, tracking and ads.  
	[WEB]	May break cross-domain logins, site functionality.
	[BUG]	May break "Cache API" when "Temporary containers" Add-on used [2], [3], [TEST].
	[1] 	https://www.ghacks.net/2017/11/22/how-to-enable-first-party-isolation-in-firefox/
	[2]	https://github.com/stoically/temporary-containers/issues/393
	[3]	https://github.com/ghacksuserjs/ghacks-user.js/issues/925
	[TEST]	https://firefox-storage-test.glitch.me/     */
user_pref("privacy.firstparty.isolate", true);					// gHacks // [FF51+]
```

I definitely would like keep FPI Enabled. This is one of the very few options I found useful in the FF. (And it's apeear buggy also!) Others are not much interesting for me, because I'm prefer Add-ons to Mozilla's crap (They are good enough only implemening telemtry or nonsense options). And, yeah, FPI is not Mozilla's development, it comes from the TOR.

Anyway here the list of techs that FPI protects:

Cookies, Cache, HTTP Authentication, DOM Storage, Flash cookies, SSL and TLS session resumption, Shared Workers, Blob URIs, SPDY and HTTP/2,  Automated Cross-Origin Redirects, window.name, Auto-Form Fill, HSTS and HPKP Supercookies, Broadcast channels, OCSP, Favicons, Mediasource URIs and Mediastream, Speculative and Prefetched Connections.

-------------------------------------------------------------

1) Please help to interpret the result obtained. 
2) What to do next with that ???

**Comment by iWARR** [comment]
### EASIER WAY TO REPRODUCE THE "TC" BUG

You even don't need to Install/Uninstall "Temporary Containers" Addon.
Please, backup (export) your TC settings before you start.

**TEST-1**

1) First Party Isolation (FPI) = [ Enabled ]

`"privacy.firstparty.isolate" = true`

2) Make `Cache API` healthy (clean the Local storage damage): 

a) Turn OFF the checkbox: TC Options -> General -> [  ] Automatic mode
b) TC Options -> Advanced -> Reset storage -> [Wipe local storage and reset it to default]

3) Restart FF

4) Turn ON the checkbox: TC Options -> General -> [•] Automatic mode

5) [TEST]	https://firefox-storage-test.glitch.me/

**Result:** `Cache API : Bad`

-------------------------

**TEST-2**

1) First Party Isolation (FPI) = [ Disabled ]

`"privacy.firstparty.isolate" = false`

2) Make **Cache API** healthy (clean the Local storage damage): 

a) Turn OFF the checkbox: TC Options -> General -> [  ] Automatic mode
b) TC Options -> Advanced -> Reset storage -> [Wipe local storage and reset it to default]

3) Restart FF

4) Turn ON the checkbox: TC Options -> General -> [•] Automatic mode

5) [TEST]	https://firefox-storage-test.glitch.me/

**Result:** `Cache API : Good`

**Comment by iWARR** [comment]
@Thorin-Oakenpants 
Please, be so kind to reopen the [ISSUE]. I'm realy need help.

**Comment by iWARR** [comment]
I know that FPI is enabled in the "ghacks-user.js" template (in my "user.js" it was also Enabled) and some users are not affected. On the other hand, I found this one pref from the ~1000 (mine + gHacks) summary. Addons were also tested/switched. So, it's highly valuable finding for my problem (of course, the diff is reproduceable).

**Comment by Thorin-Oakenpants** [comment]
The issue doesn't need to be open for you to continue to ask questions.

But here's the thing: you're using your **own heavily modified** profile, and no-one but you can troubleshoot that. Only **you** can solve the problem. I for one am not going to get involved, unless it's narrowed down and maybe I have some insight - I've unsubscribed from this issue.

I suggest you create a new profile and start using that. Clean un-corrupted files. Add your extensions and make sure they work. Then add your pref changes in small stages.

IF you use FPI, then I suggest you set that on first, before adding extensions or site permissions etc

**Comment by crssi** [comment]
> I suggest you create a new profile and start using that. 

^^ this

**Comment by Gitoffthelawn** [comment]
@iWARR If you're using TC in Automatic Mode, why do you need FPI?  Why not just create a new tab when you go to a different site?

**Comment by iWARR** [comment]
@Thorin-Oakenpants 
OK. Understood.

@Gitoffthelawn 
>If you're using TC in Automatic Mode, why do you need FPI?
Because FPI have a wider protection specialization. See my note about "techs that FPI protects" above. They are both needed, regardless all of these about isolation.

>Why not just create a new tab when you go to a different site?
I do. This is by design.

------------------------

I've done the new tests on completely new profile. Results are the same.
In case now I have TC's author answer, I'll post some words at the TC thread.

**Comment by Thorin-Oakenpants** [comment]
https://github.com/ghacksuserjs/ghacks-user.js/issues/925#issuecomment-613178214
> They've already killed part of this [appCache] off but left "applicationCache" in window (can't be arsed finding the info right now, and it's irrelevant)

Still can't be arsed .. but I just fell over the issue while checking something else: so FYI
- [1237782](https://bugzilla.mozilla.org/show_bug.cgi?id=1237782) - FF71+ appCache in Nightly/Beta no longer supported
- [1619673](https://bugzilla.mozilla.org/show_bug.cgi?id=1619673) - FF77+ ride the train to stable - landed six minutes ago (I'm on the pulse you guys: fucking nailing it!!)

**Comment by iWARR** [comment]
Thanks, **Thorin**, good catch, as usual!

Now we sure the Cache API will be recycled into a trash box. FF44-FF77 is a pretty long train, isn't it?
If the devs hadn't wasted their time implementing nonsence, the ServiceWorkers would have been done long time ago. Two versions forward to carry on the Cache API garbge... too long...

Perhaps SW will play a more significant role in the future. Let's see. People don't like SW much, it's likely a "dark horse" (unclear and complicate thing).

----------------
So now it's time to wait until this part of the code will be kicked out. Anyway, I had/have Cache API -relative things disabled.

I'm more disturbed with slowdowns after starting the browser. Probably, this is not relative to a cache. But who knows. So many old / unresolved / wontfix issues for years and years (including Local storage).
